src/error.rs

```diff
@@ -73,16 +73,16 @@ pub struct InvalidEscapeSequence {
 	/// This points to the associated backslash character in the source text.
 	pub position: usize,
 
-	/// The byte value of the invalid escape sequence.
-	pub character: Option<u8>,
+	/// The character of the invalid escape sequence.
+	pub character: Option<char>,
```

Comment:
~~What about `substitute_bytes()`? It could have source that's not valid UTF-8. We could encounter a byte in the range 128-255, which is not a valid `char` value.~~

Nevermind, apparently those are valid codepoints.

Reply:
> It could have source that's not valid UTF-8.

It will become a [`REPLACEMENT CHARACTER`](https://doc.rust-lang.org/std/primitive.char.html#associatedconstant.REPLACEMENT_CHARACTER) as per [`bstr`](https://docs.rs/bstr/latest/bstr/#handling-of-invalid-utf-8)'s handling of invalid UTF-8.

And now that I think of it, I forgot to add a test for that:

```rs
let source = b"Hello \\\xE2";
let_assert!(Err(e) = substitute_bytes(source, &map));
assert!(e.to_string() == r"Invalid escape sequence: \�");
```

Reply:
Hm,, the numerical value is preserved if you simply pass the byte value to `char::from_u32`. For byte input I think that makes more sense than converting to the unicode replacement character. `&[u8]` is not required to be UTF-8, so for `substitute_bytes()` we shouldn't make that assumption.

Reply:
Ah, good point! I was too focused on the "display error message to user" part and missed the idea that programs may want to inspect `UnexpectedCharacter` too. I will try implementing something similar to the `u32` in #8 then.

By the way, I'm wondering, what is the designed use case for `substitute_bytes()` anyway?

Reply:
I think it's fine to use `char`. All `u8` values are actually valid codepoints, even if they arent valid UTF-8.

The main usecase for `substitute_bytes()` is processing raw files without doing UTF-8 "decoding".

For a simple tool that runs on a file and expands placeholders, we don't really need it to be valid UTF-8. We only need the placeholders and escape sequences to be ASCII.

Reply:
> All `u8` values are actually valid codepoints, even if they arent valid UTF-8.

`char` has the strongly attached semantics of [Unicode scalar value](https://doc.rust-lang.org/std/primitive.char.html) though, so it won't really be more correct or convenient compared to `u32`. I think the most "correct" way would be `enum { Unicode(char), Byte(u8) }`, but it feels a bit pedantic given it's unclear practical advantage over `u32`.

I'm thinking of something like this:

```rs
if let Some(c) = char::from_u32(unexpected) {
   write!(f, "Unexpected character: {}", c)
} else if unexpected < 128u32 && !(unexpected as u8).is_ascii_control() {
   // Just print it if it's printable.
   write!(f, "Unexpected character: {}", unexpected as char)
} else {
   // Hex-codes for non-UTF8s that aren't printable in ASCII either.
   write!(f, "Unexpected character: 0x{:02X}", unexpected)
}
```

> For a simple tool that runs on a file and expands placeholders, we don't really need it to be valid UTF-8.

I see, good point! I really wish it'd be easier to implement though:joy:

Reply:
We'll never get into the `else if` or `else` branch though. All 8 bit values are valid Unicode scalar values and can be represented by `char`:

> A char is a ‘[Unicode scalar value](https://www.unicode.org/glossary/#unicode_scalar_value)’, which is any ‘[Unicode code point](https://www.unicode.org/glossary/#code_point)’ other than a [surrogate code point](https://www.unicode.org/glossary/#surrogate_code_point). This has a fixed numerical definition: code points are in the range 0 to 0x10FFFF, inclusive. Surrogate code points, used by UTF-16, are in the range 0xD800 to 0xDFFF.

Reply:
Whoops, you're right, I was confusing UTF-8 encoded bytes with Unicode scalar values :sweat_smile:. Using `char` should be fine then.

Wait, I just realized, if a byte in the range of `128..256` got treated as a `char`, `e.character.len_utf8()` will report 2 instead of 1. We may have to use an enum after all…

Reply:
Heh, also a good point... I guess an `enum` is the way..

Note that I'm definitely on board with fixing UTF-8 bugs. But I also want to stay familiar with the implementation, so a big overhaul like this is a little scary. That means that I will still need to do a more in-depth review of the implementation changes. So far I only looked at the error module.